Add support for custom URLs to WASM binaries

sebastinez opened the pull request:
To allow users to provide a `URL` to a locally hosted oniguruma WASM binary, this commit allows users to provide a function that returns a URL object to a different WASM oniguruma binary, it should .

The options object can receive a `getOnigurumaUrlFetch` and a `getOnigurumaUrlFs` property for browser and node. 

Closes #8

wooorm commented on readme.md:
```diff
@@ -192,16 +192,22 @@ Yields:
 This is async to facilitate async loading and registering, which is currently
-only used for WASM.
+only used for WASM.  
+Use the optional `Options` object to pass functions that resolve to a `URL`
+pointing to a `vscode-oniguruma` WASM binary.
+
+*Note: The URL must result in a version 1 `onig.wasm` from `vscode-oniguruma`*
```
```suggestion
```

Better to document `Options` and `GetOnigurumaUrl` separately.
Perhaps right before `` ### `starryNight.flagToScope(flag)` ``, in this section, as a h4?

I’d recommend something along the lines of:

```markdown
#### `Options`

Configuration (optional).

###### `options.getOnigurumaUrlFetch`

Get a URL to the oniguruma WASM, typically used in browsers (`GetOnigurumaUrl`,
optional).

###### `options.getOnigurumaUrlFs`

Get a URL to the oniguruma WASM, typically used in Node.js (`GetOnigurumaUrl`,
optional).

#### `GetOnigurumaUrl`

Function to get a URL to the oniguruma WASM.

> 👉 **Note**: this must currently result in a version 1 URL of
> `onig.wasm` from `vscode-oniguruma`.

###### Returns

URL object to a WASM binary (`URL` or `Promise<URL>`).
```

sebastinez replied:
added in 63950af014610c308790850465a65168d84ac5e0

wooorm commented on readme.md:
```diff
@@ -192,16 +192,22 @@ Yields:
 
 *   `grammars` (`Array<Grammar>`)
     — grammars to support
+*   `options` (`Options`)
+    — `getOnigurumaUrlFs` for node.js w/o fetch and `getOnigurumaUrlFetch` for browser and node v18+ environments.
```
```suggestion
*   `options` (`Options`, optional)
    — configuration
```

sebastinez replied:
changed in 63950af014610c308790850465a65168d84ac5e0

wooorm reviewed (commented):
Thank you for working on this!!!

The `## Types` section in the readme should include `Options` and `GetOnigurumaUrl` too.

These changes need to be tested. I think it’s fine to only test the `*Fs` version. I think it’s fine to return a bogus URL there and catch the error.

Optionally, you can add an example. It can be quite short likely. To show your use case: self-hosting `onig.wasm` on your own website.

---

One gnarly part is still that for this project, updating `vscode-oniguruma`, is likely unrelated to semver versioning, as it is internal, which means that people that use these features, will be broken at a point in the future: https://github.com/wooorm/starry-night/issues/8#issuecomment-1344340102. I’d like to discuss this further. Do you have ideas on how to handle that?
Even if *we* would cut a major on a `vscode-oniguruma`, and explain it to users, `vscode-oniguruma` might also change internals in patches/minors, which *we* pull in, but if people have a static file on a server, there will be people that forget to update, whose websites will break.

sebastinez commented on lib/get-oniguruma.fetch.js:
```diff
@@ -1,7 +1,17 @@
 /* eslint-env browser */
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/get-oniguruma.fetch.js:
```diff
@@ -1,7 +1,17 @@
+ * Browser (and Node 18+) WASM loader.
+ *
+ * @typedef {import('./index.js').Options} Options
+ * @param {Options} [options]
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/get-oniguruma.fetch.js:
```diff
@@ -1,7 +1,17 @@
+  let url = new URL('https://esm.sh/vscode-oniguruma@1/release/onig.wasm')
+  if (options && typeof options.getOnigurumaUrlFetch === 'function') {
+    url = await options.getOnigurumaUrlFetch()
+  }
```
Yeah I also prefer the ternary way, I had some issue with jsdoc protesting, but it seems that it's now ok with it, so 63950af014610c308790850465a65168d84ac5e0 fixes this

sebastinez commented on lib/get-oniguruma.js:
```diff
@@ -1,8 +1,19 @@
 import fs from 'node:fs/promises'
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/get-oniguruma.js:
```diff
@@ -1,8 +1,19 @@
+ * Node w/o fetch.
+ *
+ * @typedef {import('./index.js').Options} Options
+ * @param {Options} [options]
```
done in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/get-oniguruma.js:
```diff
@@ -1,8 +1,19 @@
+
+  if (options && typeof options.getOnigurumaUrlFs === 'function') {
+    url = await options.getOnigurumaUrlFs()
+  }
```
same as the browser fetch function, done in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/index.js:
```diff
@@ -50,6 +50,22 @@
  * @typedef {Record<number|string, Rule>} Captures
+ *
+ * @typedef Options
+ * Configuration (optional)
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/index.js:
```diff
@@ -50,6 +50,22 @@
+ *   Get a URL to the oniguruma WASM, typically used in Node.js.
+ *
+ * @callback GetOnigurumaUrl
+ *   Get a Url to the oniguruma WASM.
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/index.js:
```diff
@@ -50,6 +50,22 @@
+ * @callback GetOnigurumaUrl
+ *   Get a Url to the oniguruma WASM.
+ *
+ *   > 👉 **Note**: this must currently result in a version 1 binary of
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/index.js:
```diff
@@ -50,6 +50,22 @@
+ *   > `onig.wasm` from `vscode-oniguruma`.
+ * @returns {URL|Promise<URL>}
+ *   URL object to a WASM binary.
+ *
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented on lib/index.js:
```diff
@@ -66,15 +82,17 @@ import {theme} from './theme.js'
  * @param {Array<Grammar>} grammars
  *   Grammars to support.
+ * @param {Options} [options]
+ *   Configuration (optional).
```
fixed in 63950af014610c308790850465a65168d84ac5e0

sebastinez commented:
> Thank you for working on this!!!

My pleasure, looking forward to using it on my own.

> The `## Types` section in the readme should include `Options` and `GetOnigurumaUrl` too.

Have listed them inline with the `Grammar` and `Root`

> These changes need to be tested. I think it’s fine to only test the `*Fs` version. I think it’s fine to return a bogus URL there and catch the error.

Added to the `test.js` file, as you explained

> Optionally, you can add an example. It can be quite short likely. To show your use case: self-hosting `onig.wasm` on your own website.

Added in e922348

> One gnarly part is still that for this project, updating `vscode-oniguruma`, is likely unrelated to semver versioning, as it is internal, which means that people that use these features, will be broken at a point in the future: [#8 (comment)](https://github.com/wooorm/starry-night/issues/8#issuecomment-1344340102). I’d like to discuss this further. Do you have ideas on how to handle that? Even if _we_ would cut a major on a `vscode-oniguruma`, and explain it to users, `vscode-oniguruma` might also change internals in patches/minors, which _we_ pull in, but if people have a static file on a server, there will be people that forget to update, whose websites will break.

Yeah I thought about that, but I'm not sure I have a good answer, I think for now we should probably have a bigger Disclaimer in the readme and maybe in the jsdoc comments, so people know they can shoot themselves in the foot by using it.. 
Maybe if it shows that more people use it, and start have issues with it _we_ can revert this PR.

wooorm commented on readme.md:
```diff
@@ -192,16 +192,18 @@ Yields:
 Create a `StarryNight` that can highlight things based on the given `grammars`.
 This is async to facilitate async loading and registering, which is currently
-only used for WASM.
+only used for WASM.  
```
```suggestion
only used for WASM.
```

sebastinez replied:
63eb1bd

wooorm commented on readme.md:
```diff
@@ -192,16 +192,18 @@ Yields:
 *   `grammars` (`Array<Grammar>`)
     — grammars to support
+*   `options` (`Options`)
+    — `configuration
```
```suggestion
    — configuration
```

sebastinez replied:
Thanks for the catch! 63eb1bd

wooorm commented on readme.md:
```diff
@@ -249,6 +251,41 @@ Yields:
+###### Returns
+
+URL object to a WASM binary (`URL` or `Promise<URL>`).
+
+
```
```suggestion

```

sebastinez replied:
63eb1bd

wooorm commented on readme.md:
````diff
@@ -249,6 +251,41 @@ Yields:
+URL object to a WASM binary (`URL` or `Promise<URL>`).
+
+
+###### Example 
+```js
````
````suggestion
###### Example 

```js
````

sebastinez replied:
63eb1bd

wooorm commented on readme.md:
````diff
@@ -249,6 +251,41 @@ Yields:
+```js
+import {createStarryNight, common} from '@wooorm/starry-night'
+
+const starryNight = await createStarryNight(common, {
+  getOnigurumaUrlFetch: () => new URL(window.location.origin + "onig.wasm")
+})
````
```suggestion
const starryNight = await createStarryNight(common, {
  getOnigurumaUrlFetch() {
    return new URL('/onig.wasm', window.location.href)
  }
})
```

a) sticks with code style used in this project (no semicolons, single quotes, no unneeded arrow functions)
b) I think it’s slightly better practise to resolve from the whole current URL, and then in the relative url resolve from it, instead of string concatenation, particularly because I think you’re missing a `/` in the concatenation?

sebastinez replied:
yeah you're right, there was a `/` missing, fixed it 63eb1bd

wooorm commented on readme.md:
```diff
@@ -1389,7 +1426,7 @@ Changes should go to upstream repos and [`languages.yml`][languages-yml] in
 
 This package is fully typed with [TypeScript][].
-It exports additional `Grammar` and `Root` types that model their respective
+It exports additional `Grammar`, `Root`, `Options` and `GetOnigurumaUrl` types that model their respective
 interfaces.
```
```suggestion
It exports additional `Grammar`, `Root`, `Options`, and `GetOnigurumaUrl`
types that model their respective interfaces.
```

Oxford comma is used throughout all docs.
Lines are wrapped at 80.

sebastinez replied:
done with 63eb1bd

wooorm commented on test.js:
```diff
@@ -35,6 +35,18 @@ test('.flagToScope(flag)', async () => {
+        getOnigurumaUrlFs: () => new URL('file:///foo/baz/onig.wasm')
+      })
+    },
+    /no such file or directory, open '\/foo\/baz\/onig.wasm'/,
```
```suggestion
    /no such file or directory/,
```

In my experience, Node frequently shows different internal C calls in their error messages, across versions and platforms. Better to not include them.

sebastinez replied:
Ok no problem, 63eb1bd removes the path part

wooorm commented on test.js:
```diff
@@ -35,6 +35,18 @@ test('.flagToScope(flag)', async () => {
+      })
+    },
+    /no such file or directory, open '\/foo\/baz\/onig.wasm'/,
+    "should throw when file at given URL can't be found"
```
```suggestion
    'should support `getOnigurumaUrlFs`'
```

Please adhere to the code style used in the project: running `npm install` and `npm test` should check and format these things and make sure everything works as expected.

sebastinez replied:
63eb1bd complies with the code style

wooorm commented:
@sebastinez Btw, if you go to the “Files changed” tab, you can click each suggestion and add them to a batch, and then commit them in one go, for the ones you agree with!

sebastinez commented:
> @sebastinez Btw, if you go to the “Files changed” tab, you can click each suggestion and add them to a batch, and then commit them in one go, for the ones you agree with!

Thanks @wooorm I prefer to handle it locally in my working copy and then push the final version.

sebastinez commented:
It seems that the tests run now fine, and the linter doesn't comply.
I also added a **Disclaimer** that talks about eventual breaking if the used static WASM differs from the one pulled in from the CDN

If you are ok with it, please lmk and I will squash the commits down to one

sebastinez commented:
Hmm ok nevermind CI says otherwise 😓 will look into them

wooorm commented on readme.md:
```diff
@@ -249,6 +251,45 @@ Yields:
+> `onig.wasm` from `vscode-oniguruma`.  
+> ☝️ **Disclaimer**: Using this functionality may break your implementation
+if the static WASM you point to differs
+from the one released by `vscode-oniguruma`
```
```suggestion

> ⚠️ **Danger**: when you use this functionality, your project might break at
> any time (when reinstalling dependencies), except when you make sure that
> the WASM binary you load manually is what our internally used
> `vscode-oniguruma` dependency expects.
> To solve this, you could for example use an npm script called `dependencies`
> (which runs everytime `node_modules` is changed) which copies
> `vscode-oniguruma/release/onig.wasm` to the place you want to host it.
```

How about this?
a) A break, so there are two paragraphs, two block quotes. Otherwise they run together
b) Stronger emoji/label
c) Offers a solution

sebastinez replied:
yeah sure, that works for me 👍

wooorm commented on readme.md:
```diff
@@ -249,6 +251,51 @@ Yields:
+Function to get a URL to the oniguruma WASM.
+
+> 👉 **Note**: this must currently result in a version 1 URL of
+> `onig.wasm` from `vscode-oniguruma`.\
```
```suggestion
> `onig.wasm` from `vscode-oniguruma`.
```

This is no longer needed: hard breaks at the end even show up as a slash

wooorm commented:
Thanks for your terrific work, released in [1.5.0](https://github.com/wooorm/starry-night/releases/tag/1.5.0)!

sebastinez commented:
Thanks @wooorm for your help and support in merging this PR! Really appreciate your work 💪🙏
